easypy/contexts.py

```diff
@@ -16,3 +20,85 @@ def enter_context(self, *key):
     def exit_context(self, *key):
         self.contexts_dict[key].pop(-1).__exit__(None, None, None)
 
+
+class _BetterGeneratorContextManager(_GeneratorContextManager):
+    "This help can handle generators and other context-managers"
+
+    def __call__(self, func):
+        if isgeneratorfunction(func):
+            def inner(*args, **kwds):
+                with self._recreate_cm():
+                    yield from func(*args, **kwds)
+        elif is_contextmanager(func):
+            @contextmanager
+            def inner(*args, **kwds):
+                with self._recreate_cm():
+                    with func(*args, **kwds) as ret:
+                        yield ret
+        else:
+            def inner(*args, **kwds):
+                with self._recreate_cm():
+                    return func(*args, **kwds)
+        return wraps(func)(inner)
+
+
+if signature(_GeneratorContextManager).parameters['args'].kind is Parameter.VAR_POSITIONAL:
```

Comment:
Would have been less atrocious to monkey-patch `_GeneratorContextManager`...

At least add some comments that you do this to deal with internal changes in Python's standard library...